File: Fanficly/Search/TagResolver.swift
import Foundation

/// Resolves user-typed tags to AO3's canonical names via AO3's autocomplete
/// endpoint, so a search matches the same works whether it comes from the
/// Search tab or the Browse filters. e.g. "Draco/Hermione" →
/// "Hermione Granger/Draco Malfoy".
/// Process-lifetime memo of `term → canonical` so re-applying the same filters
/// (or sharing a tag across Search and Browse) doesn't re-hit AO3's throttled
/// autocomplete. Canonical tag names are stable, so this never goes stale.
private actor ResolutionCache {
    static let shared = ResolutionCache()
    private var store: [String: String] = [:]
    func value(for key: String) -> String? { store[key] }
    func set(_ value: String, for key: String) { store[key] = value }
    func clear() { store.removeAll() }
}

enum TagResolver {
    /// - Parameter resolveFandoms: Browse passes `false` — its fandom is picked
    ///   from AO3's own fandom list and is already canonical, so resolving it is
    ///   a wasted round-trip (and risks matching a different fandom).
    static func resolve(
        _ filters: AO3SearchFilters,
        using client: any AO3ClientProtocol,
        resolveFandoms: Bool = true
    ) async -> AO3SearchFilters {
        var resolved = filters
        resolved.relationshipNames = await resolveList(filters.relationshipNames, field: .relationship, client: client)
        resolved.characterNames    = await resolveList(filters.characterNames,    field: .character,    client: client)
        resolved.freeformNames     = await resolveList(filters.freeformNames,     field: .freeform,     client: client)
        if resolveFandoms {
            resolved.fandomNames   = await resolveList(filters.fandomNames,       field: .fandom,       client: client)
        }
        return resolved
    }

    /// Clears the process-lifetime resolution cache. For tests, so a memoized
    /// result from one case can't bleed into another.
    static func clearCacheForTesting() async {
        await ResolutionCache.shared.clear()
    }

    private static func resolveList(_ terms: [String], field: AO3AutocompleteField, client: any AO3ClientProtocol) async -> [String] {
        var out: [String] = []
        for term in terms {
            out.append(await resolveOne(term, field: field, client: client))
        }
        return out
    }

    private static func resolveOne(_ term: String, field: AO3AutocompleteField, client: any AO3ClientProtocol) async -> String {
        let cacheKey = "\(field.rawValue)|\(term.lowercased())"
        if let cached = await ResolutionCache.shared.value(for: cacheKey) {
            return cached
        }
        let result = await resolveUncached(term, field: field, client: client)
        await ResolutionCache.shared.set(result, for: cacheKey)
        return result
    }

    private static func resolveUncached(_ term: String, field: AO3AutocompleteField, client: any AO3ClientProtocol) async -> String {
        // Try the term as typed, then a few simple fallbacks.
        for candidate in candidates(for: term, field: field) {
            if let matches = try? await client.autocomplete(field: field, term: candidate),
               let best = bestMatch(for: term, in: matches, field: field) {
                return best
            }
        }
        // Relationship couldn't be matched directly (AO3's autocomplete needs
        // the canonical full names, e.g. "Hermione Granger/Draco Malfoy", not
        // "hermione/draco"). Resolve each side to a canonical character, then
        // look up the ship.
        if field == .relationship, term.contains("/") {
            let parts = term.split(separator: "/").map { $0.trimmingCharacters(in: .whitespaces) }
            if parts.count == 2 {
                let a = await resolveOne(parts[0], field: .character, client: client)
                let b = await resolveOne(parts[1], field: .character, client: client)
                // Try the ship in both orders / separators.
                for query in ["\(a)/\(b)", "\(a) \(b)", "\(b)/\(a)", "\(b) \(a)"] {
                    if let matches = try? await client.autocomplete(field: .relationship, term: query),
                       let best = bestMatch(for: "\(a)/\(b)", in: matches, field: .relationship) {
                        return best
                    }
                }
                // Last resort: construct it from the resolved names.
                if a != parts[0] || b != parts[1] { return "\(a)/\(b)" }
            }
        }
        return term
    }

    static func candidates(for term: String, field: AO3AutocompleteField) -> [String] {
        var list = [term]
        if field == .relationship, term.contains("/") {
            let spaced = term.replacingOccurrences(of: "/", with: " ")
            list.append(spaced)
            // Also try the reversed order ("draco hermione").
            let parts = term.split(separator: "/").map { $0.trimmingCharacters(in: .whitespaces) }
            if parts.count == 2 {
                list.append("\(parts[1]) \(parts[0])")
            }
        }
        return list
    }

    /// Prefer a case-insensitive exact match, else the top suggestion,
    /// else keep the original term.
    static func bestMatch(for term: String, in matches: [String], field: AO3AutocompleteField = .freeform) -> String? {
        guard !matches.isEmpty else { return nil }
        if let exact = matches.first(where: { $0.caseInsensitiveCompare(term) == .orderedSame }) {
            return exact
        }
        if field == .relationship, term.contains("/") {
            return bestRelationshipMatch(for: term, in: matches) ?? matches.first
        }
        if field == .freeform {
            return bestFreeformMatch(for: term, in: matches)
        }
        return matches.first
    }

    /// Freeform (additional) tags are user-authored descriptors, and AO3's
    /// autocomplete ranks suggestions by popularity — so a plain tag like
    /// "Caveman" gets outranked by a more specific compound tag that merely
    /// contains it ("Caveman Derek Hale"). Blindly taking the top suggestion
    /// therefore silently rewrites the user's tag into a different, narrower one.
    /// Only accept a suggestion that is the *same* tag up to case / spacing /
    /// punctuation (so we still canonicalize e.g. "hurt comfort" → "Hurt/Comfort"
    /// or fix casing); otherwise keep exactly what the user typed.
    private static func bestFreeformMatch(for term: String, in matches: [String]) -> String {
        let typed = normalizedTag(term)
        return matches.first(where: { normalizedTag($0) == typed }) ?? term
    }

    /// Lowercased, alphanumerics only — collapses case, whitespace and
    /// punctuation differences so superficial formatting variants compare equal.
    private static func normalizedTag(_ s: String) -> String {
        s.lowercased().filter { $0.isLetter || $0.isNumber }
    }

    /// Score each candidate ship against the typed term. Penalize candidates
    /// whose `/`-arity doesn't match (a 2-person query shouldn't resolve to a
    /// 3+person ship), and reward candidates where each typed side appears as a
    /// whole word in the corresponding side of the candidate.
    /// e.g. "Bella/Edward" → prefer "Bella Swan/Edward Cullen" over
    /// "Oswald Cobblepot/Isabella/Edward Nygma" (3 parts; "Bella" only matches
    /// inside "Isabella").
    private static func bestRelationshipMatch(for term: String, in matches: [String]) -> String? {
        let typed = term.split(separator: "/").map { $0.trimmingCharacters(in: .whitespaces).lowercased() }
        guard typed.count >= 2 else { return nil }
        var best: (score: Int, value: String)?
        for candidate in matches {
            let parts = candidate.split(separator: "/").map { $0.trimmingCharacters(in: .whitespaces) }
            var score = 0
            if parts.count == typed.count { score += 10 }
            else { score -= 5 * abs(parts.count - typed.count) }
            // Each typed side should be a whole-word match in *some* part.
            // Slightly prefer same-order alignment.
            for (i, side) in typed.enumerated() {
                if i < parts.count, containsWholeWord(side, in: parts[i].lowercased()) {
                    score += 3
                } else if parts.contains(where: { containsWholeWord(side, in: $0.lowercased()) }) {
                    score += 2
                } else {
                    score -= 2
                }
            }
            if best == nil || score > best!.score {
                best = (score, candidate)
            }
        }
        return best?.value
    }

    private static func containsWholeWord(_ needle: String, in haystack: String) -> Bool {
        guard !needle.isEmpty else { return false }
        let tokens = haystack.split { !$0.isLetter && !$0.isNumber }.map(String.init)
        return tokens.contains(needle)
    }
}

File: android/app/src/main/java/io/github/yennster/fanficly/search/TagResolver.kt
package io.github.yennster.fanficly.search

import io.github.yennster.fanficly.model.AO3SearchFilters
import io.github.yennster.fanficly.model.AutocompleteField
import io.github.yennster.fanficly.net.AO3Client
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Resolves user-typed tags to AO3's canonical names via AO3's autocomplete
 * endpoint, so a search matches the same works whether it comes from the smart
 * search box or the (eventual) Browse filters — e.g. `Draco/Hermione` →
 * `Hermione Granger/Draco Malfoy`. The Kotlin port of the iOS `TagResolver`.
 *
 * Results are memoized for the process lifetime (canonical names are stable, so
 * this never goes stale) to avoid re-hitting the throttled autocomplete when the
 * same tag is searched again.
 */
object TagResolver {

    private val cache = HashMap<String, String>()
    private val cacheLock = Mutex()

    /**
     * @param resolveFandoms pass `false` when the fandom came from AO3's own
     *   fandom list and is already canonical (resolving it is a wasted round-trip
     *   and risks matching a different fandom).
     */
    suspend fun resolve(
        filters: AO3SearchFilters,
        client: AO3Client,
        resolveFandoms: Boolean = true,
    ): AO3SearchFilters {
        val resolved = filters.copy()
        resolved.relationshipNames = resolveList(filters.relationshipNames, AutocompleteField.RELATIONSHIP, client)
        resolved.characterNames = resolveList(filters.characterNames, AutocompleteField.CHARACTER, client)
        resolved.freeformNames = resolveList(filters.freeformNames, AutocompleteField.FREEFORM, client)
        if (resolveFandoms) {
            resolved.fandomNames = resolveList(filters.fandomNames, AutocompleteField.FANDOM, client)
        }
        return resolved
    }

    /** Clears the process-lifetime resolution cache (for tests). */
    suspend fun clearCacheForTesting() {
        cacheLock.withLock { cache.clear() }
    }

    private suspend fun resolveList(
        terms: List<String>,
        field: AutocompleteField,
        client: AO3Client,
    ): List<String> = terms.map { resolveOne(it, field, client) }

    private suspend fun resolveOne(term: String, field: AutocompleteField, client: AO3Client): String {
        val key = "${field.path}|${term.lowercase()}"
        cacheLock.withLock { cache[key] }?.let { return it }
        val result = resolveUncached(term, field, client)
        cacheLock.withLock { cache[key] = result }
        return result
    }

    private suspend fun resolveUncached(term: String, field: AutocompleteField, client: AO3Client): String {
        // Try the term as typed, then a few simple fallbacks.
        for (candidate in candidates(term, field)) {
            val matches = runCatching { client.autocomplete(field, candidate) }.getOrNull()
            if (matches != null) {
                bestMatch(term, matches, field)?.let { return it }
            }
        }
        // A relationship couldn't be matched directly (AO3's autocomplete needs
        // the canonical full names, e.g. "Hermione Granger/Draco Malfoy", not
        // "hermione/draco"). Resolve each side to a canonical character, then
        // look up the ship.
        if (field == AutocompleteField.RELATIONSHIP && term.contains("/")) {
            val parts = term.split("/").map { it.trim() }
            if (parts.size == 2) {
                val a = resolveOne(parts[0], AutocompleteField.CHARACTER, client)
                val b = resolveOne(parts[1], AutocompleteField.CHARACTER, client)
                for (query in listOf("$a/$b", "$a $b", "$b/$a", "$b $a")) {
                    val matches = runCatching { client.autocomplete(AutocompleteField.RELATIONSHIP, query) }.getOrNull()
                    if (matches != null) {
                        bestMatch("$a/$b", matches, AutocompleteField.RELATIONSHIP)?.let { return it }
                    }
                }
                // Last resort: construct it from the resolved names.
                if (a != parts[0] || b != parts[1]) return "$a/$b"
            }
        }
        return term
    }

    /** Visible for testing. */
    fun candidates(term: String, field: AutocompleteField): List<String> {
        val list = mutableListOf(term)
        if (field == AutocompleteField.RELATIONSHIP && term.contains("/")) {
            list += term.replace("/", " ")
            val parts = term.split("/").map { it.trim() }
            if (parts.size == 2) list += "${parts[1]} ${parts[0]}"
        }
        return list
    }

    /**
     * Prefer a case-insensitive exact match, else the best-scored ship (for
     * relationships), else the top suggestion. Visible for testing.
     */
    fun bestMatch(term: String, matches: List<String>, field: AutocompleteField = AutocompleteField.FREEFORM): String? {
        if (matches.isEmpty()) return null
        matches.firstOrNull { it.equals(term, ignoreCase = true) }?.let { return it }
        if (field == AutocompleteField.RELATIONSHIP && term.contains("/")) {
            return bestRelationshipMatch(term, matches) ?: matches.first()
        }
        return matches.first()
    }

    /**
     * Score each candidate ship against the typed term. Penalize candidates whose
     * `/`-arity doesn't match (a 2-person query shouldn't resolve to a 3+person
     * ship), and reward candidates where each typed side appears as a whole word
     * in the corresponding side of the candidate.
     */
    private fun bestRelationshipMatch(term: String, matches: List<String>): String? {
        val typed = term.split("/").map { it.trim().lowercase() }
        if (typed.size < 2) return null
        var best: Pair<Int, String>? = null
        for (candidate in matches) {
            val parts = candidate.split("/").map { it.trim() }
            var score = 0
            if (parts.size == typed.size) score += 10 else score -= 5 * kotlin.math.abs(parts.size - typed.size)
            for ((i, side) in typed.withIndex()) {
                if (i < parts.size && containsWholeWord(side, parts[i].lowercase())) {
                    score += 3
                } else if (parts.any { containsWholeWord(side, it.lowercase()) }) {
                    score += 2
                } else {
                    score -= 2
                }
            }
            if (best == null || score > best.first) best = score to candidate
        }
        return best?.second
    }

    private fun containsWholeWord(needle: String, haystack: String): Boolean {
        if (needle.isEmpty()) return false
        val tokens = haystack.split(Regex("[^\\p{L}\\p{N}]+")).filter { it.isNotEmpty() }
        return tokens.contains(needle)
    }
}
